docs/contributing/effects-tutorial.md

```diff
@@ -0,0 +1,634 @@
+    Just value -> do
+      -- Check if the value existed previously
+      isNewValue <- gets (Set.member value)
+      -- We've seen this value before
+      modify (Set.insert value)
+      if isNewValue
+        then do
+          -- This is a new value
+          tell [value]
+          liftIO sleepForOneSecond
+          doTheProgram
+        else doTheProgram
+```
+
+`mtl`-style monad transformers are quite fast, compared to other, more generic
+solutions, and are on par with the non-`mtl` version.  The _reason_ for this is
+that the `m` in the return value of `doTheProgram` is generic at the definition
+site, but when we call it with an explicit list of monad runners in our `main`
+function, that generic `m` is monomorphized to
+`ReaderT AppConfig (StateT (Set ValidValue) (WriterT [ValidValue] IO))`.
+Remember, generics are slow until they're monomorphized.
+
+Now let's do it in `fused-effects`:
+
+```haskell
+main :: IO ()
+main = do
+  conf <- readConfig
+  -- We gained proper argument ordering, but lost type inference.
+  values <- execWriter @[ValidValue] $ evalState @SeenValues mempty $ runReader conf doTheProgram
+  traverse_ (putStrLn . takeValue) values
+
+doTheProgram ::
+  ( Has (Reader AppConfig) sig m
+  , Has (Writer [ValidValue]) sig m
+  , Has (State SeenValues) sig m
+  , Has (Lift IO) sig m
+  ) =>
+  m ()
+doTheProgram = do
+  -- Get the config from the reader
+  filepath <- asks takeConfig
+  -- Read the actual file
+  contents <- sendIO $ readFile filepath
+  case extractValue contents of
+    -- Value is invalid, terminate
+    Nothing -> pure ()
+    Just value -> do
+      -- Check if the value existed previously
+      isNewValue <- gets (Set.member value)
+      -- We've seen this value before
+      modify (Set.insert value)
+      if isNewValue
+        then do
+          -- This is a new value
+          tell [value]
+          sendIO sleepForOneSecond
+          doTheProgram
+        else doTheProgram
+```
+
+Note that other than some different function names (`modify'` -> `modify` and
+`liftIO` -> `sendIO`), and a very different (but still similar) style of listing
+function constraints, `doTheProgram` didn't change at all.  Our main slightly
+changed, but only such that we had to be specific about types. 
+
+Again, we can see that our `m` in `doTheProgram` is generic at the call site,
+but in our main function, we can see that it monomorphizes to
+`ReaderC AppConfig (StateC SeenValues (WriterC [ValidValue] IO))`.
+That's almost identical to our monad transformer monomorphization, and that
+similarity is very intentional.  Much of our intution from monad transformers
+is transferrable to `fused-effects`-style monads, and vice versa.
+
+## Pieces of the Puzzle
+
+Let's look at the pieces of `fused-effects`, and compare their purposes. We'll
+use the `Reader` concept as an example:
+
+| Name | Purpose |
+| --- | --- |
+| Effects | GADT-style datatypes which serve as message formats for an abstract interface. |
```

Comment:
Should we add GADTs to the list of things a reader should be familiar with?  I suppose the examples don't touch on them. Maybe a short definition of some less-important terms in the introduction?